lib/puppet/provider/service/base.rb

```diff
@@ -32,15 +32,28 @@ def getpid
     @resource.fail "Either stop/status commands or a pattern must be specified" unless @resource[:pattern]
     regex = Regexp.new(@resource[:pattern])
     ps = getps
+
     self.debug "Executing '#{ps}'"
-    IO.popen(ps) { |table|
-      table.each_line { |line|
-        if regex.match(line)
-          self.debug "Process matched: #{line}"
-          ary = line.sub(/^\s+/, '').split(/\s+/)
-          return ary[1]
-        end
-      }
+    table = Puppet::Util::Execution.execute(ps)
+
+    # The output of the PS command can be a mashup of several different
+    # encodings depending on which processes are running and what
+    # arbitrary data has been used to set their name in the process table.
+    #
+    # First, try a polite conversion to in order to match the UTF-8 encoding
+    # of our regular expression.
+    table = Puppet::Util::CharacterEncoding.convert_to_utf_8(table)
+    # If that fails, force to UTF-8 and then scrub as most uses are scanning
+    # for ACII-compatible program names.
+    table.force_encoding(Encoding::UTF_8) unless table.encoding == Encoding::UTF_8
+    table = Puppet::Util::CharacterEncoding.scrub(table) unless table.valid_encoding?
```

Comment:
I don't think that will work. The problem that this patch is trying to solve is that the output of ps can be in a **mixed encoding** as each process running on the system may have decided to write arbitrary bytes into their individual /proc/cmdline entries. In the case of the ticket that sparked this pull request it looked like raw binary data was being dumped into the command line --- not even something that had a meaningful text encoding.

Given that Puppet manifests can only express UTF-8 regexes at the moment, scrubbing the entire output block to UTF-8 after trying a conversion from Encoding.default_external seems like the best way to go.